### PR TITLE
Alerting/testing promql extraction

### PR DESCRIPTION
**Pull request by owen-d**
This introduces some custom unmarshaling for instant queries (I can't believe we didn't have this elsewhere, feel free to point me in the right direction), which the testing route then uses to map query responses into the data types expected. This achieves parity for both grafana managed & lotex based queries using grafana's internal dataframes.

Vector extraction:
```
$ curl -s admin:admin@localhost:3000/api/v1/rule/test/35 -XPOST -H 'Content-Type: application/json' -d '{ "expr": "sum by (namespace) (agent_prometheus_active_instances) >  0" }' | jq '.'
{
  "instances": [
    {
      "schema": {
        "name": "evaluation results",
        "fields": [
          {
            "name": "namespace",
            "type": "string",
            "typeInfo": {
              "frame": "string"
            }
          },
          {
            "name": "State",
            "type": "string",
            "typeInfo": {
              "frame": "string"
            }
          }
        ]
      },
      "data": {
        "values": [
          [
            "gem-perf",
            "default"
          ],
          [
            "Alerting",
            "Alerting"
          ]
        ]
      }
    }
  ]
}
```

Scalar extraction:
```
$ curl -s admin:admin@localhost:3000/api/v1/rule/test/35 -XPOST -H 'Content-Type: application/json' -d '{ "expr": "1+1"}' | jq '.'
{
  "instances": [
    {
      "schema": {
        "name": "evaluation results",
        "fields": [
          {
            "name": "State",
            "type": "string",
            "typeInfo": {
              "frame": "string"
            }
          }
        ]
      },
      "data": {
        "values": [
          [
            "Alerting"
          ]
        ]
      }
    }
  ]
}
```

### CODE REVIEW

**Review by gotjosh — APPROVED** [review]
LGTM. Great job 🚀

File: pkg/services/ngalert/api/promql_compat.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	s.V = f
+	return nil
+}
+
+func (d *queryData) UnmarshalJSON(b []byte) error {
+	type plain queryData
+	if err := json.Unmarshal(b, (*plain)(d)); err != nil {
+		return err
+	}
+
```
Comment (gotjosh) [inline pkg/services/ngalert/api/promql_compat.go:67]:
```suggestion
// Scalars and Vectors are the query result formats supported by Prometheus-like interfaces.
// https://prometheus.io/docs/prometheus/latest/querying/api/#expression-query-result-formats
```

Comment (gotjosh) [reply pkg/services/ngalert/api/promql_compat.go:67]: Mostly for posterity.

Comment (owen-d) [reply pkg/services/ngalert/api/promql_compat.go:67]: oops missed this :(

**Comment by torkelo** [comment]
@owen-d this is missing backport label (but maybe is intentional)

**Pull request merged**